beluga/include/beluga/algorithm/amcl_core.hpp

```diff
@@ -0,0 +1,237 @@
+
+ public:
+  /// Construct a AMCL instance.
+  /**
+   * \param motion_model Motion model instance.
+   * \param sensor_model Sensor model Instance.
+   * \param random_state_generator A callable able to produce random states, optionally based on the current particles
+   * state.
+   * \param spatial_hasher A spatial hasher instance capable of computing a hash out of a particle state.
+   * \param params Parameters for AMCL implementation.
+   * \param execution_policy Policy to use when processing particles.
+   */
+  Amcl(
+      MotionModel motion_model,
+      SensorModel sensor_model,
+      RandomStateGenerator random_state_generator,
+      spatial_hasher_type spatial_hasher,
+      const AmclParams& params = AmclParams{},
+      ExecutionPolicy execution_policy = std::execution::seq)
+      : params_{params},
+        motion_model_{std::move(motion_model)},
+        sensor_model_{std::move(sensor_model)},
+        execution_policy_{std::move(execution_policy)},
+        spatial_hasher_{std::move(spatial_hasher)},
+        random_probability_estimator_{params_.alpha_slow, params_.alpha_fast},
+        update_policy_{beluga::policies::on_motion(params_.update_min_d, params_.update_min_a)},
+        resample_policy_{beluga::policies::every_n(params_.resample_interval)},
+        random_state_generator_(std::move(random_state_generator)) {
+    if (params_.selective_resampling) {
+      resample_policy_ = resample_policy_ && beluga::policies::on_effective_size_drop;
+    }
+  }
+
+  /// Returns a reference to the current set of particles.
+  [[nodiscard]] const auto& particles() const { return particles_; }
+
+  /// Initialize particles using a custom distribution.
+  template <class Distribution>
+  void initialize(Distribution distribution) {
+    particles_ = beluga::views::sample(std::move(distribution)) |                    //
+                 ranges::views::transform(beluga::make_from_state<particle_type>) |  //
+                 ranges::views::take_exactly(params_.max_particles) |                //
+                 ranges::to<beluga::TupleVector>;
+    force_update_ = true;
+  }
+
+  /// Initialize particles with a given pose and covariance.
+  /**
+   * \tparam CovarianceT type representing a covariance, compliant with state_type.
+   * \throw std::runtime_error If the provided covariance is invalid.
+   */
+  template <class CovarianceT>
+  void initialize(state_type pose, CovarianceT covariance) {
+    initialize(beluga::MultivariateNormalDistribution{pose, covariance});
+  }
+
+  /// Update the map used for localization.
+  void update_map(map_type map) { sensor_model_.update_map(std::move(map)); }
+
+  /// Update particles based on motion and sensor information.
+  /**
+   * This method performs a particle filter update step using motion and sensor data. It evaluates whether
+   * an update is necessary based on the configured update policy and the force_update flag. If an update
```

Comment:
@serraramiro1 nit^2: consider 
```suggestion
   * an update is necessary based on the configured update policy. If an update is required or if it was forced via `force_update()`, ...
```
instead of discussing private state.

Reply:
Done!